### PR TITLE
[Python-experimental] JSON schema 'null' type should be modeled as 'none_type'

### PR DESCRIPTION
**Pull request by sebastien-rosset**
It looks like recently something changed because the JSON schema `null` type is now causing problems. I'm not sure what specific commit caused this problem. The `getTypeString` function in PythonClientExperimentalCodegen.java is incorrectly returning `null` when the input Schema is the Null type.

For example, with the following schema, the generated python-experimental code has syntax errors. 

```yaml
Foo:
  oneOf:
    - type: 'null'
    - '#/components/schemas/Bar'
```

Without this PR, the following import is incorrectly generated:
```python
try:
    from openapi_client.models import null
except ImportError:
    null = sys.modules[
        'openapi_client.models.null']
```

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

**Comment by sebastien-rosset** [comment]
@spacether , I'm not sure if I should handle the OAS `null` type in the `getTypeString` function (as proposed in this PR), or by adding the "null" type in the `typeMapping` field.

**Comment by spacether** [comment]
I think that we need to add the type mapping but I have not used that before. For Python-experimental we should point it to our already defined none_type.

**Comment by sebastien-rosset** [comment]
> I think that we need to add the type mapping but I have not used that before. For Python-experimental we should point it to our already defined none_type.

ok, sure, let me try with typeMappings.

File: samples/openapi3/client/petstore/python-experimental/petstore_api/models/fruit_req.py [hunk 1]
```diff
@@ -39,6 +39,11 @@
 except ImportError:
     banana_req = sys.modules[
         'petstore_api.models.banana_req']
+try:
```
Comment (sebastien-rosset) [inline samples/openapi3/client/petstore/python-experimental/petstore_api/models/fruit_req.py:42]: I need to fix this such that this import is not generated.

File: samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml [hunk 2]
```diff
@@ -1973,6 +1973,7 @@ components:
           type: string
     fruitReq:
       oneOf:
+      - type: "null"
```
Comment (sebastien-rosset) [inline samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml:1976]: I've just added this "null" type here to have one scenario with the "null" type in OneOf. @spacether , are you ok with this?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java [hunk 3]
```diff
@@ -388,7 +388,9 @@ private void fixModelImports(Set<String> imports) {
                 composedSchemaSets.add(cm.oneOf);
                 for (Set<String> importSet : composedSchemaSets) {
                     for (String otherModelName : importSet) {
-                        cm.imports.add(otherModelName);
+                        if (!"none_type".equals(otherModelName)) {
```
Comment (sebastien-rosset) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:391]: @spacether , this is one way to ensure the import below is not generated, but it's brittle. Is there a better way to skip these special case imports?

```python
try:
    from openapi_client.models import null
except ImportError:
    null = sys.modules[
        'openapi_client.models.null']
```

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java [hunk 4]
```diff
@@ -388,7 +388,9 @@ private void fixModelImports(Set<String> imports) {
                 composedSchemaSets.add(cm.oneOf);
                 for (Set<String> importSet : composedSchemaSets) {
                     for (String otherModelName : importSet) {
-                        cm.imports.add(otherModelName);
+                        if (!languageSpecificPrimitives.contains(otherModelName)) {
```
Comment (sebastien-rosset) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:391]: This is to avoid creating an import for "none_type".

File: samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml (hunk 2 above)
Comment (spacether) [reply samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml:1976]: Yup, I am okay adding this too our sample spec

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache [hunk 5]
```diff
@@ -1043,6 +1050,13 @@ def get_anyof_instances(self, model_args, constant_args):
         return anyof_instances
 
     for anyof_class in self._composed_schemas()['anyOf']:
+        if anyof_class is none_type:
+            # The only way the input data can match the 'None' type
+            # is if the input data is the 'null' value.
+            if model_args is None:
+                # The input data matches the 'None' type.
+                oneof_instances.append(model_args)
```
Comment (spacether) [inline modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1058]: Did you mean anyof_instances.append..?
Also, I don't think that we should add model_args to anyof_instances.
anyof_instances is meant to store ModelSimple or ModelComposed instances, not None.
Let's just delete this if clause and continue.

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache [hunk 6]
```diff
@@ -1043,6 +1050,13 @@ def get_anyof_instances(self, model_args, constant_args):
         return anyof_instances
 
     for anyof_class in self._composed_schemas()['anyOf']:
+        if anyof_class is none_type:
+            # The only way the input data can match the 'None' type
+            # is if the input data is the 'null' value.
+            if model_args is None:
```
Comment (spacether) [inline modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1056]: If model_args can be None please update our docstrings to show that
- here +
- get_oneof_instance
- any calling function that passes it in

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache (hunk 5 above)
Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1058]: Yes, good catch, I copy/pasted.

Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1058]: > Did you mean anyof_instances.append..?
> Also, I don't think that we should add model_args to anyof_instances.
> anyof_instances is meant to store ModelSimple or ModelComposed instances, not None.
> Let's just delete this if clause and continue.

But then potentially `if len(anyof_instances) == 0:` would be true, if the input data is the `null` value and the OAS schema has `anyOf: [type: 'null', a, b, c]`

Comment (spacether) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache]: Yes, return None. The spec says that None is a valid value. We raise exceptions if deserialization fails.

Comment (spacether) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache]: We never get to counting oneOf or anyOf instance because we returned None in the `ComposedSchema.__new__` method and `ComposedSchema.__init__` is never called.

Comment (spacether) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache]:
```
a = FruitReq(None)
a == None -> True
isisnstance(a, FruitReq) -> False
isisnstance(a, none_type) -> True
```

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache (hunk 6 above)
Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1060]: done

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache (hunk 5 above)
Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache]: done

**Comment by sebastien-rosset** [comment]
> Please add a python test showing deserializing a None working correctly.

Added unit test

File: modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml [hunk 7]
```diff
@@ -1855,6 +1855,7 @@ components:
           - $ref: '#/components/schemas/banana'
     fruitReq:
        oneOf:
+          - type: 'null'
```
Comment (spacether) [inline modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml:1858]: Can you add this to the v3 python-experimental spec also?

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache [hunk 8]
```diff
@@ -1024,9 +1024,10 @@ def get_oneof_instance(self, model_args, constant_args):
     schemas described by oneOf.
     Args:
         self: the class we are handling
-        model_args (dict): var_name to var_value
+        model_args (dict/None): var_name to var_value
```
Comment (spacether) [inline modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1027]: With the neswest code we will never reach here when deserializing None.
Can you remove the None additions from this docstirng?
- this also applies to get_anyof_instances

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache [hunk 9]
```diff
@@ -1041,6 +1042,13 @@ def get_oneof_instance(self, model_args, constant_args):
     # Iterate over each oneOf schema and determine if the input data
     # matches the oneOf schemas.
     for oneof_class in self._composed_schemas['oneOf']:
+        if oneof_class is none_type:
+            # The only way the input data can match the 'None' type
+            # is if the input data is the 'null' value.
+            if model_args is None:
+                # The input data matches the 'None' type.
+                oneof_instances.append(model_args)
```
Comment (spacether) [inline modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1050]: Can you delete this `if model_args is None:` block?
We will still go through this code path for deserializing pojo payloads but we should continue through the None like you are doing.
- this also applies to get_anyof_instances

File: samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py [hunk 10]
```diff
@@ -125,4 +125,16 @@ def test_deserialize_mammal(self):
...
+        """
+        deserialize fruit
+        fruitReq is a oneOf composed schema model with discriminator, including 'null' type.
+        """
+
+        # Unmarshal 'null' value
+        data = None
+        response = MockResponse(data=json.dumps(data))
+        deserialized = self.deserialize(response, (petstore_api.FruitReq, type(None)), True)
+        self.assertEqual(type(deserialized), type(None))
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: I don't think that this will work anymore.
How about [here](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache#L66)
adding:
```
oneof_anyof_classes = cls._composed_schemas.get('oneOf', ()) + cls._composed_schemas.get('anyOf', ())
if oneof_anyof_classes and none_type in oneof_anyof_classes and args[0] is None:
  return None
```
Then you can also delete the oneof_anyof_classes definition from lines 78-80.

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache (hunk 9 above)
Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache]: done

File: modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml (hunk 7 above)
Comment (sebastien-rosset) [reply modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml:1858]: done

File: samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py (hunk 10 above)
Comment (spacether) [reply samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: Was this done; where? In the PR I don't see any code added in the __new__ method.

Comment (sebastien-rosset) [reply samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: > Where was this done? In the PR I don't see any code added in the `__new__` method.

Where was what done?

Comment (spacether) [reply samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: My above suggestion asked that you add code in the new method. So I am asking where the requested code was added.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java [hunk 11]
```diff
@@ -2986,6 +2993,12 @@ public CodegenProperty fromProperty(String name, Schema p) {
         } else if (ModelUtils.isFreeFormObject(p)) {
             property.isFreeFormObject = true;
         } else if (ModelUtils.isAnyTypeSchema(p)) {
+            // The 'null' value is allowed when the OAS schema is 'any type'.
+            // See https://github.com/OAI/OpenAPI-Specification/issues/1389
+            if (Boolean.FALSE.equals(p.getNullable())) {
+                LOGGER.error("Schema '{}' is any type, which includes the 'null' value. 'nullable' cannot be set to 'false'", p.getName());
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:2999]: I would use LOGGER.warn instead and make it clear it's for the upcoming OAS v3.1

(after OAS v3.1 release, I would expect the parser (swagger-parser) to catch this kind of issue)

Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: done

File: samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py (hunk 10 above)
Comment (sebastien-rosset) [reply samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: ok, I missed that proposed change.

Comment (sebastien-rosset) [reply samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: Is it expected that your proposed change has no check for "allOf"?

Comment (spacether) [reply samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: Can you describe a payload and schema that would need this allOf check where the check is missing?

when deserializing None where none is in OneOf or AnyOf, any allOf definition of object property fields can never be met by a None payload because None is not a pojo.

Further down in the new function we cover the use caas where we have allof definitionsl, verify that the payload conforms to it then deserialize to an anyof or oneof pojo payload.

Comment (sebastien-rosset) [reply samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: That was a question for you. You wrote this code, so I am asking you.

Comment (spacether) [reply samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: Does my above answer meet your needs? I don't understand when we would need to check allof but we miss that check. In the answer I describe both cases that do and don't check allof and why.

Comment (sebastien-rosset) [reply samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py:140]: Thanks. I just copied your code as is without checking the correctness.

**Comment by spacether** [comment]
@sebastien-rosset we need the CI tests to pass before we can merge this.
Travis says:
```
E     File "/home/travis/build/OpenAPITools/openapi-generator/samples/client/petstore/python-experimental/petstore_api/model_utils.py", line 137
E       oneof_anyof_classes = cls._composed_schemas.get('oneOf', ()) +
```
It might need parens around that statement because it is multiline, like:
```
oneof_anyof_classes = (first_part +
  second_part)
```
Can you fix it?

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache [hunk 12]
```diff
@@ -1041,6 +1043,16 @@ def get_oneof_instance(self, model_args, constant_args):
     # Iterate over each oneOf schema and determine if the input data
     # matches the oneOf schemas.
     for oneof_class in self._composed_schemas['oneOf']:
+        # The composed oneOf schema allows the 'null' type and the input data
+        # is the null value. This is a OAS >= 3.1 feature.
+        if oneof_class is none_type:
```
Comment (spacether) [inline modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1048]: We will only reach this code if we are deserializing a dict payload.
This is because the `__new__` method is run before the `__init__` method. And in `__new__` we return None rather than an instance of the composed class when we pass in None and oneOf contains None type.
So in that case, get_oneof_instance is never called. So in this loop, all we need to do is skip over the none_types.
Can you change this if block to be only this?
```
if oneof_class is none_type:
  # skip none_types because we are deserializing dict data. none_type deserialization is handled in the __new__ method
  continue
```

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache [hunk 13]
```diff
@@ -1098,6 +1112,16 @@ def get_anyof_instances(self, model_args, constant_args):
         return anyof_instances
 
     for anyof_class in self._composed_schemas['anyOf']:
+        # The composed oneOf schema allows the 'null' type and the input data
+        # is the null value. This is a OAS >= 3.1 feature.
+        if anyof_class is none_type:
```
Comment (spacether) [inline modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1117]: We will only reach this code if we are deserializing a dict payload.
This is because the `__new__` method is run before the `__init__` method. And in `__new__` we return None rather than an instance of the composed class when we pass in None and oneOf contains None type.
So in that case, get_anyof_instances is never called. So in this loop, all we need to do is skip over the none_types.
Can you change this if block to be only this?
```
if anyof_class is none_type:
  # skip none_types because we are deserializing dict data. none_type deserialization is handled in the __new__ method
  continue
```

File: modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache (hunk 12 above)
Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1048]: Actually, I just added this code because otherwise the existing unit tests fail. This code IS executed when the unit tests are executed.

Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache:1048]: Thank you for the change. I just tested it. It works.

**Comment by spacether** [comment]
@sebastien-rosset to fix the CI-error you need to update this line:
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/python-experimental/model_templates/model_simple.mustache#L14
to:
```
_composed_schemas = {}
```

**Comment by sebastien-rosset** [comment]
> @sebastien-rosset to fix this next CI-error you need to update this line:
> https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/python-experimental/model_templates/model_simple.mustache#L14
> to:
> 
> ```
> _composed_schemas = {}
> ```
> 
> or change the oneof_anyof_classes assignment in `__new__` to only assign a value if bool(cls._composed_schemas) == True which is equivalent to:
> 
> ```
> oneof_anyof_classes = ()
> if cls._composed_schemas:
>   oneof_anyof_classes = (cls._composed_schemas.get('oneOf', ()) +
>                                 cls._composed_schemas.get('anyOf', ()))
> ```

I have some something like that here: https://github.com/OpenAPITools/openapi-generator/pull/5809/files#diff-cc72ac40056827e3a103f9b4f56059cfR70

I mentioned #5809 should be merged before this PR.
